agent/internal/client/base_client.go

```diff
@@ -95,3 +99,30 @@ func (c *baseCommunicator) resetClient() {
 		c.cedarHTTPClient.Timeout = 0
 	}
 }
+
+func (c *baseCommunicator) createCedarGRPCConn(ctx context.Context, comm Communicator) error {
+	if c.cedarGRPCClient == nil {
+		cc, err := comm.GetCedarConfig(ctx)
+		if err != nil {
+			return errors.Wrap(err, "getting cedar config")
+		}
+
+		dialOpts := timber.DialCedarOptions{
+			BaseAddress: cc.BaseURL,
+			RPCPort:     cc.RPCPort,
+			Username:    cc.Username,
+			APIKey:      cc.APIKey,
+			Retries:     10,
+		}
+		c.cedarGRPCClient, err = timber.DialCedar(ctx, c.cedarHTTPClient, dialOpts)
+		if err != nil {
+			return errors.Wrap(err, "creating cedar grpc client connection")
+		}
+	}
+
+	// We should always check the health of the conn as a sanity check,
+	// this way we can fail the agent early and avoid task system failures.
+	healthClient := gopb.NewHealthClient(c.cedarGRPCClient)
+	_, err := healthClient.Check(ctx, &gopb.HealthCheckRequest{})
+	return errors.Wrap(err, "checking cedar grpc health")
```

Comment:
Should this attempt to re-connect if the Cedar GPRC client is no longer valid (e.g. the connection is invalidated due to Cedar restarting)?

Reply:
The connection client takes care of that. From https://pkg.go.dev/google.golang.org/grpc#ClientConn:
> A ClientConn encapsulates a range of functionality including name resolution, TCP connection establishment (with retries and backoff) and TLS handshakes. It also handles errors on established connections by re-resolving the name and reconnecting.